tests/js/server/aql/aql-traversal-paths-optimization.js

```diff
@@ -0,0 +1,216 @@
+///
+/// @author Jan Steemann
+/// @author Copyright 2020, triAGENS GmbH, Cologne, Germany
+////////////////////////////////////////////////////////////////////////////////
+
+const jsunity = require("jsunity");
+const db = require("@arangodb").db;
+const { deriveTestSuite } = require('@arangodb/test-helper');
+const isCoordinator = require('@arangodb/cluster').isCoordinator();
+const isEnterprise = require("internal").isEnterprise();
+  
+const vn = "UnitTestsVertex";
+const en = "UnitTestsEdges";
+const gn = "UnitTestsGraph";
+
+function BaseTestConfig() {
+  return {
+
+    testProducePathsOutputs: function () {
+      [
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN p", [ true, true, false ]], 
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN p[0]", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN LENGTH(p)", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN p.edges", [ false, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN p.vertices", [ true, false, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN [p.edges, p.vertices]", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN [p.edges, p]", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN [p.vertices, p]", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN [p.edges, p.vertices, p]", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' RETURN p.testi", [ true, false, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' COLLECT pl = LENGTH(p) RETURN pl", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' COLLECT key = v._key INTO g RETURN { key, other: g[*].p.edges }", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' FILTER LENGTH(p) > 0 RETURN v", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' SORT p.edges RETURN v", [ false, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' SORT p.vertices RETURN v", [ true, false, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' SORT p.edges, p.vertices RETURN v", [ true, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' SORT LENGTH(p.edges) RETURN v", [ false, true, false ]],
+        [ "FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' SORT LENGTH(p.vertices) RETURN v", [ true, false, false ]],
```

Comment:
For the sake of completeness we should also include "wierd" ways to write edges / vertices:

```
RETURN p.`vertices`
```
```
RETURN p["vertices"]
```

```
RETURN p[CONCAT("vert","ices")]
```

And also make the access not single attribute:
```
RETURN p.vertices[*].name
```

// Maybe not even array access, this will return null btw.
```
RETURN p.vertices.name
```

I do expect our optimizer to handle all of the weird writing methods before hand, just to be on the save side here.

Also something that comes to my mind is subqueries.
```
"FOR v, e, p IN 1..1 OUTBOUND '" + vn + "/test0' GRAPH '" + gn + "' LET sub = (FOR x IN p.vertices RETURN x.name) RETURN sub"
```